lib/rugged_adapter/git_layer_rugged.rb

```diff
@@ -361,12 +363,13 @@ def build_log(sha, options)
                 commit_touches_path?(c, current_path, options[:follow], walker)
                 # This is a commit we care about, unless we haven't skipped enough
                 # yet
+                pathnames[c.oid] = current_path.dup if track_pathnames && !pathnames.has_key?(c.oid)
                 skipped += 1
                 commits.push(Gollum::Git::Commit.new(c)) if skipped > offset
               end
           end
         walker.reset
-        commits
+        track_pathnames ? [commits, pathnames] : commits
```

Comment:
Yes, I was thinking that.

Also, `pathnames` would be empty when `options[:follow]` was not set, and I thought it could be confusing to return an empty Hash when not having specified `:follow`?

Reply:
But uncertain about what's best here!

Reply:
It's my preference to have methods always return the same kind of object or objects, and to have the caller figure out how to handle it from there. Perhaps we should also consider alternatives, such as commits knowing about (or knowing how to obtain) their pathnames.

Reply:
> commits knowing about (or knowing how to obtain) their pathnames.

Going for that option!